packages/backend-cli/src/directLine.ts

```diff
@@ -0,0 +1,772 @@
+      }
+    })
+  );
+}
+
+export class DirectLine implements IBotConnection {
+  public connectionStatus$ = new BehaviorSubject(ConnectionStatus.Uninitialized);
+  public activity$: Observable<Activity>;
+  private fetchAsObservable: (options: any) => Observable<{ response: ArrayBuffer | any }>;
+  private createFormData: (attachments: FormDataAttachment[]) => any;
+
+  private domain = "https://directline.botframework.com/v3/directline";
+  private webSocket;
+
+  private conversationId: string;
+  private secret: string;
+  private token: string;
+  private watermark = '';
+  private streamUrl: string;
+  public referenceGrammarId: string;
+
+  private pollingInterval: number = 1000;
+
+  private tokenRefreshSubscription: Subscription;
+
+  constructor(options: DirectLineOptions) {
+    this.createFormData = options.createFormData;
+    this.fetchAsObservable = fetchAsObservable.bind(this, options.fetch);
+    this.secret = options.secret;
+    this.token = options.secret || options.token;
+    this.webSocket = (options.webSocket === undefined ? true : options.webSocket) && typeof WebSocket !== 'undefined' && WebSocket !== undefined;
+
+    if (options.domain)
+      this.domain = options.domain;
+    if (options.conversationId) {
+      this.conversationId = options.conversationId;
+    }
+    if (options.watermark) {
+      if (this.webSocket)
+        console.warn("Watermark was ignored: it is not supported using websockets at the moment");
+      else
+        this.watermark =  options.watermark;
+    }
+    if (options.streamUrl) {
+      if (options.token && options.conversationId)
+        this.streamUrl = options.streamUrl;
+      else
+        console.warn("streamUrl was ignored: you need to provide a token and a conversationid");
+    }
+    if (options.pollingInterval !== undefined)
+      this.pollingInterval = options.pollingInterval;
+
+    this.activity$ = (this.webSocket
+      ? this.webSocketActivity$()
+      : this.pollingGetActivity$()
+    ).share();
+  }
+
+  // Every time we're about to make a Direct Line REST call, we call this first to see check the current connection status.
+  // Either throws an error (indicating an error state) or emits a null, indicating a (presumably) healthy connection
+  private checkConnection(once = false) {
+    let obs =  this.connectionStatus$
+    .flatMap(connectionStatus => {
+      if (connectionStatus === ConnectionStatus.Uninitialized) {
+        this.connectionStatus$.next(ConnectionStatus.Connecting);
+
+        //if token and streamUrl are defined it means reconnect has already been done. Skipping it.
+        if (this.token && this.streamUrl) {
+          this.connectionStatus$.next(ConnectionStatus.Online);
+          return Observable.of(connectionStatus);
+        } else {
+          return this.startConversation().do(conversation => {
+            this.conversationId = conversation.conversationId;
+            this.token = this.secret || conversation.token;
```

Comment:
start conversation sometimes doesn't return a token? Seems like we would prefer token over secret.